hydrate/match.py

```diff
@@ -0,0 +1,193 @@
+        self.match_categories = []
+
+        FullMatchResults = self.get_full_matches(cluster_components)
+
+        if FullMatchResults.full_matches:
+            self.match_categories.append(
+                MatchCategory(top_comment=FULL_MATCH_COMMENT,
+                              start_index=self.get_start_index(),
+                              matches=FullMatchResults.full_matches))
+
+        if FullMatchResults.leftovers:
+            PartialMatchResults = self.get_partial_matches(FullMatchResults.leftovers)
+
+            if PartialMatchResults.partial_matches:
+                self.match_categories.append(
+                    MatchCategory(top_comment=PARTIAL_MATCH_COMMENT,
+                                  start_index=self.get_start_index(),
+                                  matches=PartialMatchResults.partial_matches))
+
+            if PartialMatchResults.leftovers:
+                NoMatchResults = self.get_no_matches(PartialMatchResults.leftovers)
+
+                self.match_categories.append(
+                    MatchCategory(top_comment=NO_MATCH_COMMENT,
+                                  start_index=self.get_start_index(),
+                                  matches=NoMatchResults))
+
+        return self.match_categories
+
+    def get_start_index(self):
+        """Calculate start_index based on the last MatchCategory in the list."""
+        start_idx = 0
+        if self.match_categories:
+            start_idx = len(self.match_categories[-1].matches) \
+                        + self.match_categories[-1].start_index
+        return start_idx
+
+    def get_full_matches(self, cluster_components):
+        """Determine which components fully match the cluster.
+
+        Returns:
+            FullMatchResults: namedtuple
+                full_matches: list of components or []
+                leftovers: list of components or None
+
+        """
+        FullMatchResults = namedtuple('FullMatchResults', ['full_matches', 'leftovers'])
+        full_matches = []
+        leftovers = None
+        cluster_map = {cc.name: cc for cc in cluster_components}
+        cluster_set = set(cluster_map.keys())
+
+        for rc in self.repo_components:
+            repo_set = set(rc.name.split('-'))
+            if repo_set.issubset(cluster_set):
+                cc_names = cluster_set.intersection(repo_set)
+                cluster_matches = [cluster_map[component_name] for component_name in cc_names]
+                cluster_set.difference_update(repo_set)
+
+                full_matches.append(
+                    Match(category=MatchCategory.FULL_MATCH,
+                          cluster_matches=cluster_matches,
+                          repo_match=rc))
+
+        if cluster_set:
+            leftovers = [cc for cc in cluster_components if cc.name in cluster_set]
+
+        return FullMatchResults(full_matches, leftovers)
+
+    def get_partial_matches(self, full_match_leftovers):
```

Comment:
I feel like you can merge get_partial_matches and get_full_matches and just take the match category as argument: category=MatchCategory.FULL_MATCH
get_matches(category=MatchCategory.FULL_MATCH) have one by default if they pass partial then do the partial
those two methods look very similar

Reply:
I feel like the methods build off of each other too much to be able to specify the type of match in an argument. The partial matching method relies on the results of the full match method, so they need to be called in order. I do see what you mean that they are very similar... there's definitely repeated code.